meadow-client/src/MainFrame.purs

```diff
@@ -547,38 +556,39 @@ evalF (ApplyTransaction next) = do
     Nothing -> pure next
 
 evalF (NextBlock next) = do
-  modifying (_marloweState <<< _blockNum) (add one)
+  saveInitialState
+  modifying _marloweState (addToNEL (updateStateP <<< (over _blockNum (add one))))
   updateState
   pure next
 
 evalF (AddAnyInput {person, anyInput} next) = do
-  modifying (_marloweState <<< _transaction <<< _inputs) ((flip snoc) anyInput)
+  modifying (_marloweState <<< _Head <<< _transaction <<< _inputs) ((flip snoc) anyInput)
   case person of
     Just per -> do
-      modifying (_marloweState <<< _transaction <<< _signatures) (Map.insert per true)
+      modifying (_marloweState <<< _Head <<< _transaction <<< _signatures) (Map.insert per true)
       updateState
       pure next
     Nothing -> do
       updateState
       pure next
 
 evalF (RemoveAnyInput anyInput next) = do
-  modifying (_marloweState <<< _transaction <<< _inputs) (delete anyInput)
+  modifying (_marloweState <<< _Head <<< _transaction <<< _inputs) (delete anyInput)
   updateState
   pure next
 
 evalF (SetChoice {idChoice: (IdChoice {choice, person}), value} next) = do
-  modifying (_marloweState <<< _input <<< _choiceData) (Map.update (Just <<< (Map.update (const $ Just value) choice)) person)
+  modifying (_marloweState <<< _Head <<< _input <<< _choiceData) (Map.update (Just <<< (Map.update (const $ Just value) choice)) person)
   updateState
   pure next
 
 evalF (SetOracleVal {idOracle, value} next) = do
-  modifying (_marloweState <<< _input <<< _oracleData) (Map.update (\x -> Just (x {value = value})) idOracle)
+  modifying (_marloweState <<< _Head <<< _input <<< _oracleData) (Map.update (\x -> Just (x {value = value})) idOracle)
   updateState
   pure next
 
 evalF (SetOracleBn {idOracle, blockNumber} next) = do
-  modifying (_marloweState <<< _input <<< _oracleData) (Map.update (\x -> Just (x {blockNumber = blockNumber})) idOracle)
+  modifying (_marloweState <<< _Head <<< _input <<< _oracleData) (Map.update (\x -> Just (x {blockNumber = blockNumber})) idOracle)
```

Comment:
I think you'll be able to do:

```
assign (_marloweState <<< _Head <<< _input <<< _oracleData <<< ix idOracle <<< _blockNumber) blockNumber
```

Reply:
There are a few places this applies.

Reply:
This is the kind of nitpicking you get if you introduce whitespace changes and draw attention to old code. :-P

Reply:
_But it's not nitpicking. This is a solid gold suggestion. - Ed._

Reply:
I like this a lot

Reply:
Worth knowing that you use `ix` if you want to access a key that must be present and `at` for creating/deleting.

Reply:
In particular:
```
evalF (SetChoice {idChoice: (IdChoice {choice, person}), value} next) = do
  modifying (_currentInput <<< _choiceData) (Map.update (Just <<< (Map.update (const $ Just value) choice)) person)
  assign (_currentInput <<< _choiceData <<< ix person <<< ix choice) value
```
This is what lenses were made for!

Reply:
👏 👏 👏